optuna/storages/_rdb/storage.py

```diff
@@ -463,7 +463,46 @@ def _create_new_trial(
         """
 
         session = self.scoped_session()
+        try:
+            # Locking within a study is necessary since the creation of a trial is not an atomic
+            # operation. More precisely, the trial number computed in `_get_prepared_new_trial` is
+            # prone to race conditions within this lock.
+            session.query(models.StudyModel).filter(
+                models.StudyModel.study_id == study_id
+            ).with_for_update().one()
+
+            trial = self._get_prepared_new_trial(study_id, template_trial, session)
+            self._commit(session)
+        except OperationalError:
+            # Retry after deadlock. This may happen in distributed environments.
+            session.rollback()
+            return self._create_new_trial(study_id, template_trial)
```

Comment:
I think that we should not do retry after deadlock. If deadlock is happen, we should reconsider the logic. What do you think?

Reply:
Thanks for you comment. Could you elaborate, also explain why that is? It seemed natural to me (https://dev.mysql.com/doc/refman/5.7/en/innodb-deadlocks-handling.html) but to be frank, it's not easily reproduced (encountered it a few times with MySQL).

Reply:
Hmm. I've ever thought it's unnatural like deadlock on multi-threading programming. But according to the MySQL documentation you shared, it seems to be wrong. Thank you for sharing useful link.

> but to be frank, it's not easily reproduced (encountered it a few times with MySQL).

BTW, I mind the reason why current implementation will raise deadlock such frequently.

Reply:
> such frequently.

Just to clarify, it happens very **in**frequently.

Reply:
Oh, you are right. I misread your comment.

Reply:
Sounds no problem. Thank you for your reply.